tla-io/src/main/scala/at/forsyte/apalache/io/itf/ItfToTla.scala

```diff
@@ -0,0 +1,249 @@
+package at.forsyte.apalache.io.itf
+
+import at.forsyte.apalache.io.itf.ItfToTla._
+import at.forsyte.apalache.io.json.{JsonDeserializationError, JsonRepresentation, ScalaFactory}
+import at.forsyte.apalache.io.lir.Trace
+import at.forsyte.apalache.tla.lir._
+import at.forsyte.apalache.tla.typecomp.TBuilderInstruction
+import at.forsyte.apalache.tla.types.parser.DefaultType1Parser
+import at.forsyte.apalache.tla.types.{tla, ModelValueHandler}
+
+/**
+ * Reads a sequence of States, i.e. mappings from variable names to TlaEx values, from an ITF json.
+ *
+ * @tparam T
+ *   Any class extending JsonRepresentation
+ * @author
+ *   Jure Kukovec
+ */
+class ItfToTla[T <: JsonRepresentation](
+    scalaFactory: ScalaFactory[T]) {
+
+  /**
+   * Checks that the Apalache-specific ITF requirements are met, as defined in ADR15. If yes, returns a mapping from
+   * variables to their types.
+   */
+  private[itf] def validateShapeAndGetTypes(json: T): Map[String, TlaType1] = {
+    val varTypesOpt = for {
+      meta <- json.getFieldOpt(META_FIELD)
+      varTypes <- meta.getFieldOpt(VAR_TYPES_FIELD)
+    } yield varTypes
+    val varTypesJson = varTypesOpt.getOrElse {
+      throw new JsonDeserializationError("ITF JSON does not satisfy Apalache requirements. " +
+        "Toplevel \"#meta\" field must exist and contain \"varTypes\". " +
+        "See ADR15: https://apalache.informal.systems/docs/adr/015adr-trace.html#itf-as-input-to-apalache. ")
+    }
+
+    val diffOpt = for {
+      vars <- json.getFieldOpt(VARS_FIELD)
+      varTypesKeys <- varTypesJson.allFieldsOpt
+    } yield {
+      val varsSet = scalaFactory.asSeq(vars).map { scalaFactory.asStr }.toSet
+      // Symmetric difference
+      (varsSet -- varTypesKeys) ++ (varTypesKeys -- varsSet)
+    }
+
+    val diff = diffOpt.getOrElse {
+      throw new JsonDeserializationError(s"\"$VARS_FIELD\" or \"$VAR_TYPES_FIELD\" contents are malformed.")
```

Comment:
Seems like the logic would be clearer if you first validate whether the fields are present and in the expected form, and handle reporting errors on that, then do the validation of whether there is a type for each variable, rather than mixing up the two different validation checks here.

Reply:
My reasoning is a bit different, I try to catch every form of exception the moment there is sufficient evidence to record a violation. Also, to me, whether there is a type for every variable and whether fields are missing are not really different things; in both cases the input is equally malformed.

Reply:
> I try to catch every form of exception the moment there is sufficient evidence to record a violation

I guess you mean "catch every form of error". This is basically what I'm encouraging: you are currently performing inspection of the sets in the code "prior" to handling errors from invalid data. And, more generally, the code does not handle report the error at the earliest point it is detectable, instead requiring followup computations after other things have been processed, and wrapping several different kinds of errors into one.

Here's an example of how using `Either`s as I'm suggesting could help clean up the error handling here (imo). iiuc, this snippet accounts for all the entire body of this method:

```scala
  private[itf] def validateShapeAndGetTypes(json: T): Either[String, Map[String, TlaType1]] = for {
    meta <- json.getFieldOpt(META_FIELD).toRight("'#meta' field must exist")
    varTypes <- meta.getFieldOpt(VAR_TYPES_FIELD).toRight("'#meta' field must contain a `varTypes` field")
    varTypesKeys <- varTypes.allFieldsOpt.toRight("'#meta.varTypes' must be a map")
    vars <- json.getFieldOpt(VARS_FIELD).toRight("'vars' field must exist")
    varsSet = scalaFactory.asSeq(vars).map { scalaFactory.asStr }.toSet
    _ <- {
      val varsWithoutTypes = (varsSet -- varTypesKeys)
      Either.cond(varsWithoutTypes.isEmpty, (),
          s"$VARS_FIELD contains variables missing types in $VAR_TYPES_FIELD: ${varsWithoutTypes}")
    }
    _ <- {
      val typesWithoutVars = (varTypesKeys -- varsSet)
      Either.cond(typesWithoutVars.isEmpty, (),
          s"$VAR_TYPES_FIELD contains variables missing types in $VARS_FIELD: ${typesWithoutVars}")
    }
  } yield varTypes.allFieldsOpt.get.map { k =>
    k -> DefaultType1Parser.parseType(scalaFactory.asStr(varTypes.getFieldOpt(k).get))
  }.toMap

```

IMO, this has two benefits:

1. Users get more precise information about the source of each error.
2. Readers of the code get one logical line for each error-prone case, they can see at a glance what the error case is, rather than having to look several lines to a subsequent test to sees what exception is raised.

Once you're finished sequencing your error-prone computations, you can always drop out of the `Either` monad by throwing an exception with the error, but this approach lets us clearly isolate and declare the sequence of validation steps in a safe and, imo, quite perspicuous way.

Reply:
To clarify: I think it's worth considering this approach, but I won't block on the use  of `Either`.

I do think it may be worth being more precise about which errors arising when, but if you think there are good reasons not to, I'm open to learn.